### PR TITLE
[addons] remove not needed part and add new version check way

### PR DESCRIPTION
**Pull request by AlwinEsch**
First step of the big rework

## Description
This remove on first commit the never used GetSettings add-on function. This was thought to define addon settings by the code and not with xml, but nowhere used and becomes removed.

The other two commits add the first part of the new version way. This is thought to prevent e.g. the `GetVersion` function on PVR. There becomes everything added to one header (versions.h), reason to have in one is to read from them with cmake (on next changes in new coming request) and write them to a `addon.xml.in`.

## Motivation and Context
Big binary add-on system rework.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: addons/resource.language.en_gb/resources/strings.po [hunk 1]
```diff
@@ -15232,7 +15232,13 @@ msgctxt "#29802"
 msgid "Passthrough audio in use"
 msgstr ""
 
-#empty strings from id 29803 to 29899
+#. The label in the add-ons dll class to report about outdated add-on api version
+#: xbmc/addons/AddonDll.cpp
+msgctxt "#29803"
+msgid "Using incompatible API version for type '%s'.\nKodi API version = '%s', add-on API version '%s'\n\nPlease check for updates or contact developer of add-on."
```
Comment (MartijnKaijser) [inline addons/resource.language.en_gb/resources/strings.po]: Please do not use line breaks in strings if they are not really needed.

**Comment by notspiff** [comment]
you want me to merge up front ?

**Comment by AlwinEsch** [comment]
Want to create on all repo addons the request first, to reduce time for this changes.

File: xbmc/addons/AddonDll.cpp [hunk 2]
```diff
@@ -461,5 +398,31 @@ ADDON_STATUS CAddonDll::TransferSettings()
...
+                            kodi::addon::GetTypeName(type),
+                            kodiVersion,
+                            addonVersion);
+
+    std::string heading = StringUtils::Format("%s: %s", TranslateType(Type(), true).c_str(), Name().c_str());
+    std::string text = StringUtils::Format(g_localizeStrings.Get(29803).c_str(),
+                                           kodi::addon::GetTypeName(type),
+                                           kodiVersion,
+                                           addonVersion);
+    CGUIDialogOK::ShowAndGetInput(CVariant{heading}, CVariant{text});
```
Comment (tamland) [inline xbmc/addons/AddonDll.cpp]: Don't make gui calls from deep down in the addons system please. We shouldn't have random functions adding popous. Instead this should be in the gui code tied to some user action (or notification if it's a background task)

**Comment by garbear** [comment]
because it's an ABI change (break?) can we bumb the add-on versions? Otherwise an add-on compiled against a newer kodi will fail on an older Kodi build

**Comment by ksooo** [comment]
> can we bumb the add-on versions?

 yes, please.

**Comment by AlwinEsch** [comment]
Hi @garbear @ksooo this request is to allow it in future, currently is the add-on base interface not checked for version changes, that's the main reason to add `ADDON_GetTypeVersion` function here.

After them is in brings a old add-on a load failure, but no crashes:
bildschirmfoto vom 2017-04-27 00-46-35

The Version increase does not work for them, because the library is not loaded and his functions not available.

After them comes another request who bring the new version check way, as example see this https://github.com/FernetMenta/kodi-agile/blob/master/addons/kodi.global.main/addon.xml.in#L2. This is only the first step.

File: xbmc/addons/AddonDll.cpp (hunk 2 above)
Comment (AlwinEsch) [inline xbmc/addons/AddonDll.cpp]: Have no idea for really good alternatives, want to have them only be available there, that's why I find a user message as not so good way. As alternative can I use this:
```cpp
CEventLog::GetInstance().AddWithNotification(EventPtr(new CNotificationEvent(heading, text, EventLevel::Error)));
```
who bring:
bildschirmfoto vom 2017-04-27 02-04-08

Here how it was with them here:
bildschirmfoto vom 2017-04-27 02-09-31

File: xbmc/addons/AddonDll.cpp [hunk 3]
```diff
@@ -461,5 +400,31 @@ ADDON_STATUS CAddonDll::TransferSettings()
...
+                            Name().c_str(),
+                            kodi::addon::GetTypeName(type),
+                            kodiVersion,
+                            addonVersion);
+
+    std::string text = StringUtils::Format(g_localizeStrings.Get(29803).c_str(),
+                                           kodi::addon::GetTypeName(type),
+                                           kodiVersion,
+                                           addonVersion);
+    CEventLog::GetInstance().AddWithNotification(EventPtr(new CNotificationEvent(Name(), text, EventLevel::Error)));
```
Comment (AlwinEsch) [inline xbmc/addons/AddonDll.cpp]: @tamland have changed it to the `CEventLog::GetInstance().AddWithNotification` with:
bildschirmfoto vom 2017-04-27 02-04-08

This by start of screensaver with wrong version:
bildschirmfoto vom 2017-04-27 02-24-33

**Comment by ksooo** [comment]
Not sure I got this right. But I guess you know best what you are doing. Maybe you should at least tell the addon maintainers that merging the addon PR you opened yesterday leads to a no longer functioning addon (at least until you merged this PR), because it is ABI incompatible?

**Comment by AlwinEsch** [comment]
The function call tables (e.g. PVR) are not changed, this change "C" exported functions from the base addon functions.

Goal on end is to have only a `ADDON_Create` and `ADDON_GetTypeVersion` as exported functions, everything else becomes then handled with function tables. This allow then in future to check a several years old addon is compatible.

**Comment by AlwinEsch** [comment]
jenkins build this please

**Comment by AlwinEsch** [comment]
OK to bring in?

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/libKODI_guilib.h [hunk 4]
```diff
@@ -170,10 +170,10 @@ typedef struct CB_GUILib
 
 
 /* current ADDONGUI API version */
-#define KODI_GUILIB_API_VERSION "5.11.0"
+#define KODI_GUILIB_API_VERSION ADDON_GLOBAL_VERSION_GUI
 
 /* min. ADDONGUI API version */
-#define KODI_GUILIB_MIN_API_VERSION "5.10.0"
+#define KODI_GUILIB_MIN_API_VERSION ADDON_GLOBAL_VERSION_GUI
```
Comment (stefansaraev) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/libKODI_guilib.h]:
```
# cat /usr/share/kodi/addons/kodi.guilib/addon.xml 
<?xml version="1.0" encoding="UTF-8"?>
<addon id="kodi.guilib" version="#define KODI_GUILIB_API_VERSION ADDON_GLOBAL_VERSION_GUI" provider-name="Team-Kodi">
  <backwards-compatibility abi="#define KODI_GUILIB_MIN_API_VERSION ADDON_GLOBAL_VERSION_GUI"/>
  <requires>
    <import addon="xbmc.core" version="0.1.0"/>
  </requires>
</addon>
```

any addon that depends on guilib is probably broken now. 
https://github.com/xbmc/xbmc/blob/master/cmake/scripts/common/Macros.cmake#L632-L634 should also be changed

**Pull request merged**